File: perl_scripts/homer2meme.pl
#!/usr/bin/perl
use strict;
use warnings;
use Getopt::Long;
use IO::Uncompress::Gunzip qw($GunzipError);

my $input       = '';
my $extract     = '';
my $pseudocount = 0.01;

GetOptions(
    'i=s' => \$input,
    'e=s' => \$extract,
    'a=f' => \$pseudocount,
    'h'   => sub { usage() },
) or usage();

usage() unless $input;
die "Error: -a must be > 0.\n" unless $pseudocount > 0;

my $fh;
if ($input eq '-') {
    $fh = \*STDIN;
} elsif ($input =~ /\.gz$/) {
    $fh = IO::Uncompress::Gunzip->new($input)
        or die "Cannot open $input: $GunzipError";
} else {
    open $fh, '<', $input or die "Cannot open $input: $!";
}

my $header_printed = 0;
my $in_motif = 0;
my $motif_id = '';
my $description = '';
my @matrix;

while (<$fh>) {
    chomp;
    next unless length($_);

    # HOMER header line starts with '>'
    if (/^>(.*)/) {
        my $rest = $1;

        # Flush previous motif
        if ($in_motif && @matrix) {
            print_meme_header() unless $header_printed;
            $header_printed = 1;
            print_meme_motif($motif_id, $description, \@matrix);
        }
        @matrix = ();

        my @parts = split /\t/, $rest;
        my $mid  = defined $parts[0] ? $parts[0] : 'motif';
        my $desc = defined $parts[1] ? $parts[1] : $mid;

        # Apply -e filter
        if ($extract && $mid ne $extract && $desc ne $extract) {
            $in_motif = 0;
            next;
        }

        $motif_id    = $mid;
        $description = $desc;
        $in_motif    = 1;
        next;
    }

    next unless $in_motif;

    # Matrix data row
    my @tokens = split /\s+/;
    my $all_numeric = 1;
    for my $t (@tokens) {
        unless ($t =~ /^-?[\d.]+([eE][+-]?\d+)?$/) {
            $all_numeric = 0;
            last;
        }
    }
    next unless $all_numeric && @tokens;

    my @row = map { $_ + 0 } @tokens;
    if (scalar(@row) != 4) {
        warn "Warning: skipping malformed matrix row (expected 4 cols, got "
             . scalar(@row) . "): $_\n";
        next;
    }

    # Auto-detect log-odds vs probability (prob rows sum to ~1.0)
    my $sum = 0;
    $sum += $_ for @row;
    if ($sum < 0.98 || $sum > 1.02) {
        @row = logodds_to_prob(\@row, $pseudocount);
    }
    push @matrix, \@row;
}

# Flush last motif
if ($in_motif && @matrix) {
    print_meme_header() unless $header_printed;
    print_meme_motif($motif_id, $description, \@matrix);
}

# Close handle correctly depending on type
if ($input ne '-') {
    if (ref $fh && $fh->isa('IO::Uncompress::Gunzip')) {
        $fh->close() or warn "Error closing gz file: $GunzipError";
    } else {
        close $fh or warn "Error closing file: $!";
    }
}

# ---------------------------------------------------------------------------

sub logodds_to_prob {
    my ($row_ref, $pc) = @_;
    my $background = 0.25;
    my @raw = map { 2 ** $_ * $background } @$row_ref;
    my $total = $pc * scalar(@raw);
    $total += $_ for @raw;
    return map { ($_ + $pc) / $total } @raw;
}

sub print_meme_header {
    print "MEME version 4\n";
    print "\n";
    print "ALPHABET= ACGT\n";
    print "\n";
    print "strands: + -\n";
    print "\n";
    print "Background letter frequencies\n";
    print "A 0.25 C 0.25 G 0.25 T 0.25\n";
    print "\n";
}

sub print_meme_motif {
    my ($id, $desc, $matrix_ref) = @_;
    my $width = scalar @$matrix_ref;
    print "MOTIF $id $desc\n";
    print "\n";
    print "letter-probability matrix: alength= 4 w= $width nsites= 20 E= 0\n";
    foreach my $row (@$matrix_ref) {
        print "  " . join("  ", map { sprintf("%.6f", $_) } @$row) . "\n";
    }
    print "\n";
}

sub usage {
    print <<EOF;
Usage: $0 -i <input_file> [OPTIONS]

Convert HOMER motif format to MEME format.

Options:
    -i <file>   Input HOMER motif file (or '-' for stdin, supports .gz)
    -e <string> Extract only specified motif by id or description
    -a <float>  Pseudocount for log-odds to probability conversion (default: 0.01)
    -h          Show this help

Examples:
    $0 -i results/motifs.homer > raw/motifs.meme
    $0 -i results/motifs.homer.gz > raw/motifs.meme
    $0 -i results/motifs.homer -e "CTCF/Jaspar"
    cat motifs.homer | $0 -i -

EOF
    exit 0;
}

File: rust_scripts/src/bin/homer2meme.rs
//! homer2meme - Convert HOMER motif format to MEME motif format
//!
//! Build (Cargo, recommended - includes .gz support):
//!   cd rust_scripts && cargo build --release
//!   # Binaries: rust_scripts/target/release/meme2homer  homer2meme
//!
//! Install to PATH:
//!   cd rust_scripts && cargo install --path .
//!
//! Usage:
//!   homer2meme -i motifs.homer > motifs.meme
//!   homer2meme -i motifs.homer.gz > motifs.meme
//!   cat motifs.homer | homer2meme -i -

use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::process;
use flate2::read::MultiGzDecoder;

struct Args {
    input: String,
    extract: String,
    pseudocount: f64,
}

fn parse_args() -> Args {
    let argv: Vec<String> = env::args().collect();
    let mut input = String::new();
    let mut extract = String::new();
    let mut pseudocount = 0.01_f64;
    let mut i = 1usize;
    while i < argv.len() {
        match argv[i].as_str() {
            "-i" => {
                i += 1;
                if i >= argv.len() { eprintln!("Error: -i requires a value"); process::exit(1); }
                input = argv[i].clone();
            }
            "-e" => {
                i += 1;
                if i >= argv.len() { eprintln!("Error: -e requires a value"); process::exit(1); }
                extract = argv[i].clone();
            }
            "-a" => {
                i += 1;
                if i >= argv.len() { eprintln!("Error: -a requires a value"); process::exit(1); }
                let raw = argv[i].clone();
                pseudocount = raw.parse().unwrap_or_else(|_| {
                    eprintln!("Invalid -a value: {}", raw);
                    process::exit(1);
                });
                if pseudocount <= 0.0 {
                    eprintln!("Error: -a must be > 0, got {}", pseudocount);
                    process::exit(1);
                }
            }
            "-h" | "--help" => { usage(); process::exit(0); }
            other => { eprintln!("Unknown option: {}", other); usage(); process::exit(1); }
        }
        i += 1;
    }
    if input.is_empty() {
        eprintln!("Error: -i <input_file> is required.");
        usage();
        process::exit(1);
    }
    Args { input, extract, pseudocount }
}

fn usage() {
    eprintln!(r#"Usage: homer2meme -i <input_file> [OPTIONS]

Convert HOMER motif format to MEME format.

Options:
    -i <file>    Input HOMER motif file (.homer, .motif, or .gz), or '-' for stdin
    -e <string>  Extract only specified motif by id or description
    -a <float>   Pseudocount for log-odds -> probability conversion (default: 0.01)
    -h           Show this help

Examples:
    homer2meme -i motifs.homer > motifs.meme
    homer2meme -i motifs.homer.gz > motifs.meme
    homer2meme -i motifs.homer -e "CTCF/Jaspar"
    cat motifs.homer | homer2meme -i -
"#);
}

fn is_logodds(row: &[f64]) -> bool {
    let s: f64 = row.iter().sum();
    !(0.98..=1.02).contains(&s)
}

fn logodds_to_prob(row: &[f64], pseudocount: f64) -> Vec<f64> {
    let background = 0.25_f64;
    let raw: Vec<f64> = row.iter().map(|&v| 2.0_f64.powf(v) * background).collect();
    let total: f64 = raw.iter().sum::<f64>() + pseudocount * raw.len() as f64;
    raw.iter().map(|&v| (v + pseudocount) / total).collect()
}

fn print_meme_header() {
    println!("MEME version 4");
    println!();
    println!("ALPHABET= ACGT");
    println!();
    println!("strands: + -");
    println!();
    println!("Background letter frequencies");
    println!("A 0.25 C 0.25 G 0.25 T 0.25");
    println!();
}

fn print_meme_motif(id: &str, desc: &str, matrix: &[Vec<f64>]) {
    let width = matrix.len();
    println!("MOTIF {} {}", id, desc);
    println!();
    println!("letter-probability matrix: alength= 4 w= {} nsites= 20 E= 0", width);
    for row in matrix {
        let line: Vec<String> = row.iter().map(|v| format!("{:.6}", v)).collect();
        println!("  {}", line.join("  "));
    }
    println!();
}

fn parse_and_convert<R: BufRead>(reader: R, args: &Args) {
    let mut header_printed = false;
    let mut in_motif = false;
    let mut motif_id = String::new();
    let mut description = String::new();
    let mut matrix: Vec<Vec<f64>> = Vec::new();

    for line_result in reader.lines() {
        let line = line_result.expect("Failed to read line");
        let trimmed = line.trim();

        if trimmed.is_empty() { continue; }

        if let Some(rest) = trimmed.strip_prefix('>') {
            if in_motif && !matrix.is_empty() {
                if !header_printed { print_meme_header(); header_printed = true; }
                print_meme_motif(&motif_id, &description, &matrix);
            }
            matrix.clear();

            let parts: Vec<&str> = rest.splitn(6, '\t').collect();
            let mid = parts.first().copied().unwrap_or("motif").to_string();
            let raw_desc = parts.get(1).copied().unwrap_or("").to_string();
            let final_desc = if raw_desc.is_empty() { mid.clone() } else { raw_desc };

            if !args.extract.is_empty() && mid != args.extract && final_desc != args.extract {
                in_motif = false;
                continue;
            }
            motif_id = mid;
            description = final_desc;
            in_motif = true;
            continue;
        }

        if !in_motif { continue; }

        let row_result: Result<Vec<f64>, _> = trimmed.split_whitespace()
            .map(|s| s.parse::<f64>()).collect();
        if let Ok(mut row) = row_result {
            if !row.is_empty() {
                if row.len() != 4 {
                    eprintln!("Warning: skipping malformed matrix row with {} cols (expected 4): {}", row.len(), trimmed);
                    continue;
                }
                if is_logodds(&row) { row = logodds_to_prob(&row, args.pseudocount); }
                matrix.push(row);
            }
        }
    }

    if in_motif && !matrix.is_empty() {
        if !header_printed { print_meme_header(); }
        print_meme_motif(&motif_id, &description, &matrix);
    }
}

fn main() {
    let args = parse_args();
    if args.input == "-" {
        parse_and_convert(BufReader::new(io::stdin().lock()), &args);
    } else if args.input.ends_with(".gz") {
        let file = File::open(&args.input)
            .unwrap_or_else(|e| { eprintln!("Cannot open {}: {}", args.input, e); process::exit(1); });
        parse_and_convert(BufReader::new(MultiGzDecoder::new(file)), &args);
    } else {
        let file = File::open(&args.input)
            .unwrap_or_else(|e| { eprintln!("Cannot open {}: {}", args.input, e); process::exit(1); });
        parse_and_convert(BufReader::new(file), &args);
    }
}
